1354 dpl 928 application environment indicator

wendyyang opened the pull request:
Closes #1354 

#### Changes proposed in this pull request
separate header component;
class to header background color is based on the process.env.NODE_ENV environment;
change the footer background color to be dependent on environment variable too;  

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

seenanair commented on src/components/TractionHeader.vue:
```diff
@@ -0,0 +1,46 @@
+      },
+    }
+
+</script>
```
We are currently progressing towards composition APIs in Traction. Therefore, it might be good to have this new component as a composition one.

seenanair commented on src/components/TractionHeader.vue:
```diff
@@ -0,0 +1,46 @@
+            let bgColorClass = '';
+            if (this.environment == 'development'){
+                bgColorClass = 'bg-blue-600';
+            }else if(this.environment == 'staging'){
```
Is there no 'uat' environment or is the 'uat' environment called as 'staging'?

wendyyang replied:
In fact, I am not sure what uat environment will return.  Will it only be known when deployed in 'uat'? 
I actually guess it might just return to be 'production' in uat.

stevieing replied:
> In fact, I am not sure what uat environment will return. Will it only be known when deployed in 'uat'? I actually guess it might just return to be 'production' in uat.

We might need to add an environment specific variable that we add to the .env file and the deployment project e.g.

VITE_ENVIRONMENT = development

seenanair replied:
Yes, I think so, I think it should be uat for uat environment. @stevieing please correct if I am wrong.

wendyyang replied:
good point!

seenanair commented on src/components/InfoFooter.vue:
```diff
@@ -66,6 +66,19 @@ export default {
+                bgColorClass = 'bg-gradient-to-tr';
+            }
+            return [ 'border border-t-2 pt-5 pb-3 flex-shrink-0 border-sdb-100', bgColorClass ];
+        }
```
Just a thought - since we are accessing this across multiple places, perhaps we can move this to a common place using composables.

seenanair reviewed (commented):
Looking good 👍  I have just shared a few thoughts.

stevieing reviewed (changes requested):
A good start moving it into a separate component.

Agreed with @seenanair about using composition API and composables.

stevieing commented on src/components/InfoFooter.vue:
```diff
@@ -66,6 +66,19 @@ export default {
+            let bgColorClass = '';
+            if (this.environment == 'development'){
+                bgColorClass = 'bg-blue-600';
+            }else if(this.environment == 'staging'){
```
should this be UAT?

wendyyang commented on src/components/InfoFooter.vue:
```diff
@@ -66,6 +66,19 @@ export default {
+            }else if(this.environment == 'production'){
+                bgColorClass = 'bg-gradient-to-tr';
+            }
+            return [ 'border border-t-2 pt-5 pb-3 flex-shrink-0 border-sdb-100', bgColorClass ];
```
Thank you for your suggestion. That is a good idea.

seenanair commented on src/lib/BgColorClass.js:
```diff
@@ -0,0 +1,7 @@
+const BgColorClass = {
```
Just a thought , I think it is a common convention to use camelCase with js objects and also matches with rest of codebase

wendyyang replied:
changed.

seenanair reviewed (commented):
Just few questions
- I couldn't see any colour changes in the footer. It would be better to have the footer colour changed as well.
- Is the uat environment added in deployment as well?

wendyyang commented:
> Just few questions
> 
> * I couldn't see any colour changes in the footer. It would be better to have the footer colour changed as well.
> * Is the uat environment added in deployment as well?

I changed footer color the same way as header. In footer , there was an environment variable import.meta.env.MODE which only can be 'development' or 'production', and it displays in the footer. Now I changed the environment variable to be import.meta.env['VITE_ENVIRONMENT'], it can be 'development', 'uat' or 'production'. which means the messages displayed in footer can be one of the three. I hope it is Ok. 
the uat environment is set in .env.uat VITE_ENVIRONMENT.

seenanair commented:
> > Just few questions
> > 
> > * I couldn't see any colour changes in the footer. It would be better to have the footer colour changed as well.
> > * Is the uat environment added in deployment as well?
> 
> I changed footer color the same way as header. In footer , there was an environment variable import.meta.env.MODE which only can be 'development' or 'production', and it displays in the footer. Now I changed the environment variable to be import.meta.env['VITE_ENVIRONMENT'], it can be 'development', 'uat' or 'production'. which means the messages displayed in footer can be one of the three. I hope it is Ok. the uat environment is set in .env.uat VITE_ENVIRONMENT.

That sounds fine. I think we need to add VITE_ENVIRONMENT in sanger/deployment project as well where we need to set it appropriately so it works in uat and production environments

dasunpubudumal commented on src/components/InfoFooter.vue:
```diff
@@ -56,11 +56,13 @@
   data() {
     return {
-      environment: import.meta.env.MODE,
+      environment: import.meta.env['VITE_ENVIRONMENT'],
```
Is there a specific reason we're moving away from [`MODE`](https://vitejs.dev/guide/env-and-mode#env-variables)?

It looks like having a `VITE_ENVIRONMENT=production` in `.env.production` and `VITE_ENVIRONMENT=uat` in `.env.uat` is a bit redundant. Vite builds and [attaches the correct environment](https://vitejs.dev/guide/env-and-mode#modes) name when running: 

```bash
vite build --mode <environment name>
```

This is just me getting some clarifications :)

wendyyang replied:
Does 'import.meta.env.MODE' only return 'production' & 'development'? can't distinguish 'uat' & 'production'?

stevieing replied:
Yes. Node only has development and production.

dasunpubudumal replied:
I think it depends on the build script. So, if you build your application with `vite build --mode uat`, it would attach the configurations (at build time) in `.env.uat` (take a look at the [documentation](https://vitejs.dev/guide/env-and-mode#modes) how they've utilised `staging`).

I'm not well versed in this area. Let's bring some experts in 😄 @seenanair

seenanair replied:
Yes, for uat I think we need to specifically set the mode. But, I do not know we are using that for uat deployment.
 So either we can  change script in deployment project for uat build which allows mode to be used , which allows meta.env.MODE to be used across all environments or  to keep this VITE_ENVIRONMENT flag and manage it

dasunpubudumal replied:
(Sorry @stevieing, didn't see your reply when I was typing [this reply](https://github.com/sanger/traction-ui/pull/1762#discussion_r1660737927)).

So, are we building the UI using something other than Vite? In [`package.json`](https://github.com/sanger/traction-ui/blob/3208cb24e294df6f8b1f0ae126b634c2dbc907a9/package.json#L12) it looks like we're using `vite build` though.

stevieing replied:
We would have to make changes to the github action as we build on release not in the deployment project. Not entirely sure how we would do this at this stage as we would need to set the mode when we build a release so we would need to do it conditionally either for develop or master.

stevieing commented on .env.production:
```diff
@@ -3,4 +3,5 @@ VITE_TRACTION_GRAPHQL_URL=REPLACE_VITE_TRACTION_GRAPHQL_URL
-VITE_SEQUENCESCAPE_API_KEY=REPLACE_VITE_SEQUENCESCAPE_API_KEY 
\ No newline at end of file
+VITE_SEQUENCESCAPE_API_KEY=REPLACE_VITE_SEQUENCESCAPE_API_KEY
+VITE_ENVIRONMENT=production
```
This needs to be VITE_ENVIRONMENT=REPLACE_VITE_ENVIRONMENT and then REPLACE_VITE_ENVIRONMENT=production in the deployment project.

wendyyang replied:
Do I need to make a branch in deployment project for this change?

stevieing reviewed (changes requested):
All good. Just need to make changes for use in deployment project.

seenanair commented:
@wendyyang @stevieing One question - There is no.env.training. Don't we need that for this to work in training?

wendyyang commented:
> @wendyyang @stevieing One question - There is no.env.training. Don't we need that for this to work in training?

I deployed the branch to training (https://training.traction.psd.sanger.ac.uk/#/dashboard). 
it seems working without env.training
